test/distributed/_composable/fully_shard/test_fully_shard_mixed_precision.py

```diff
@@ -31,59 +31,160 @@
+            torch.float32,
+            torch.float32,
+            None,
+        ),
+        # (False, False, True): ('no_cast_root_no_cast_child_no_root_params',
+        #                        torch.float32, torch.float32, torch.float32),  # expected result after eval mode fix
+        (False, False, False): (
+            "no_cast_root_no_cast_child_root_params",
+            torch.float32,
+            torch.float32,
+            torch.float32,
+        ),
+    }
+
     @property
     def world_size(self):
         return 2
 
     @skip_if_lt_x_gpu(2)
-    def test_float16_on_one_submodule(self):
+    def test_float16_cast_forward(self):
         self.run_subtests(
             {
                 "cast_root_forward_inputs_submodule": [True, False],
                 "cast_forward_inputs_submodule": [True, False],
+                "use_root_no_params": [True, False],
             },
-            self._test_float16_on_one_submodule,
+            self._test_float16_cast_forward,
         )
 
-    def _test_float16_on_one_submodule(
+    def _test_float16_cast_forward(
         self,
         cast_root_forward_inputs_submodule: bool,
         cast_forward_inputs_submodule: bool,
+        use_root_no_params: bool,
+    ):
+        self.cast_root_forward_inputs_submodule = cast_root_forward_inputs_submodule
+        self.cast_forward_inputs_submodule = cast_forward_inputs_submodule
+        x, fsdp = self._input_and_model_init(use_root_no_params=use_root_no_params)
+        # self._validate_eval(x, fsdp)
+        self._backward_or_validate_error(x, fsdp)
+        self._assert_expected_dtypes(use_root_no_params)
+
+    def _backward_or_validate_error(
+        self, input: Dict[nn.Module, torch.Tensor], fsdp_model: nn.Module
     ):
-        forward_inputs: Dict[nn.Module, torch.Tensor] = {}
+        # cast_root_forward_inputs_submodule or cast_forward_inputs_submodule should be True
+        if (
+            not self.cast_root_forward_inputs_submodule
+            and not self.cast_forward_inputs_submodule
+        ):
+            with self.assertRaisesRegex(
+                RuntimeError,
+                "mat1 and mat2 must have the same dtype",
+            ):
+                fsdp_model(input).sum().backward()
+        else:
+            fsdp_model(input).sum().backward()
+
+    def _input_and_model_init(self, use_root_no_params: bool):
+        self.forward_inputs: Dict[nn.Module, torch.Tensor] = {}
         float16 = MixedPrecision(
             param_dtype=torch.float16,
-            cast_root_forward_inputs=cast_root_forward_inputs_submodule,
-            cast_forward_inputs=cast_forward_inputs_submodule,
+            cast_root_forward_inputs=self.cast_root_forward_inputs_submodule,
+            cast_forward_inputs=self.cast_forward_inputs_submodule,
         )
 
-        model = SaveForwardInputsModel(
-            forward_inputs=forward_inputs,
+        self.model = SaveForwardInputsModel(
```

Comment:
Curious: Do we want to save these as attributes to the test class? Are we concerned about polluting that if there are other tests? Or is the intention to create a new test class then?

Reply:
Good point! I went ahead and minimized the test class attributes. While having the test class attributes made the method signatures a bit cleaner, I agree the possibility of subsequent subtest pollution isn't worth that trade-off (especially since adding subtests should be clear/straightforward with this setup)